library/ceph_key.py

```diff
@@ -0,0 +1,422 @@
+    It can also list and get information about key(s)
+options:
+    cluster:
+        description:
+            - The ceph cluster name.
+        required: false
+        default: ceph
+    name:
+        description:
+            - name of the CephX key
+        required: true
+    state:
+        description:
+            - present|absent|list|update|info
+        required: true
+        choices: ['present', 'absent', 'list', 'update', 'info']
+        default: list
+    caps:
+        description:
+            - CephX key capabilities
+        required: false
+    secret:
+        description:
+            - key's secret value
+        required: false
+        default: None
+    containerized:
+        description:
+            - Weither or not this is a containerized cluster. The value is
+            assigned or not depending on how the playbook runs.
+        required: false
+        default: None
+'''
+
+EXAMPLES = '''
+- name: create cephx key
+  ceph_key:
+    name: "my_key"
+    state: present
+    caps: "{{ caps }}"
+
+- name: update cephx key
+  ceph_key:
+    name: "my_key"
+    state: update
+    caps: "{{ caps }}"
+
+- name: delete cephx key
+  ceph_key:
+    name: "my_key"
+    state: absent
+
+- name: info cephx key
+  ceph_key:
+    name: "my_key""
+    state: info
+
+- name: list cephx keys
+  ceph_key:
+    state: list
+'''
+
+RETURN = '''#  '''
+
+from ansible.module_utils.basic import AnsibleModule
+import datetime
+import os
+import struct
+import time
+import base64
+
+
+def fatal(message, module):
+    '''
+    Report a fatal error and exit
+    '''
+
+    if module:
+        module.fail_json(msg=message, rc=1)
+    else:
+        raise(Exception(message))
+
+
+def key_mode(file_path):
+    '''
+    Change mode file for a CephX key
+    Problem, how do to do on containerized deployment?
+    '''
+    os.chmod(file_path)
+
+
+def generate_secret():
+    '''
+    Generate a CephX secret
+    '''
+
+    key = os.urandom(16)
+    header = struct.pack('<hiih', 1, int(time.time()), 0, len(key))
+    secret = base64.b64encode(header + key)
+
+    return secret
+
+
+def generate_caps(cmd, type, caps):
+    '''
+    Generate CephX capabilities list
+    '''
+    cmd = cmd
+
+    for k, v in caps.iteritems():
+        if type == "ceph-authtool":
+            cmd.extend(["--cap"])
+
+        cmd.extend([k, v])
+
+    return cmd
+
+
+def generate_ceph_cmd(cluster, args, containerized):
+    '''
+    Generate 'ceph' command line to execute
+    '''
+
+    cmd = []
+
+    base_cmd = [
+        'ceph',
+        '--cluster',
+        cluster,
+        'auth',
+    ]
+
+    cmd.extend(base_cmd + args)
+
+    if containerized:
+        cmd = containerized.split() + cmd
+
+    return cmd
+
+
+def generate_ceph_authtool_cmd(cluster, name, secret, caps, containerized):
+    '''
+    Generate 'ceph-authtool' command line to execute
+    '''
+
+    file_destination = "/etc/ceph/" + cluster + "." + name + ".keyring"
```

Comment:
it is safer to use `os.path.join` here instead of concatenating the strings in this way, it is also duplicating the same path defined in `create_key`

Reply:
thanks